src/bigistore.c

```diff
@@ -978,9 +866,352 @@ bigistore_svals(PG_FUNCTION_ARGS)
+}
+
+PG_FUNCTION_INFO_V1(bigistore_delete_istore);
+Datum bigistore_delete_istore(PG_FUNCTION_ARGS)
+{
+    BigIStorePair *pairs, *delpairs;
+    BigIStore *    is    = PG_GETARG_BIGIS_COPY(0);
+    BigIStore *    isdel = PG_GETARG_BIGIS(1);
+
+    int index1 = 0, index2 = 0;
+
+    pairs    = FIRST_PAIR(is, BigIStorePair);
+    delpairs = FIRST_PAIR(isdel, BigIStorePair);
+
+    while (index1 < is->len && index2 < isdel->len)
+    {
+        if (pairs[index1].key < delpairs[index2].key)
+        {
+            ++index1;
+        }
+        else if (pairs[index1].key > delpairs[index2].key)
+        {
+            ++index2;
+        }
+        else
+        {
+            if (pairs[index1].val == delpairs[index2].val)
+            {
+                --is->len;
+                is->buflen -= bigis_pair_buf_len(pairs + index1);
+                if (is->len > index1)
+                    memmove(pairs + index1, pairs + index1 + 1, (is->len - index1) * sizeof(BigIStorePair));
+                ++index2;
+            }
+            else
+            {
+                ++index2;
+                ++index1;
+            }
+        }
+    }
+
+    SET_VARSIZE(is, ISHDRSZ + (is->len * sizeof(BigIStorePair)));
+    PG_RETURN_POINTER(is);
+}
+
+PG_FUNCTION_INFO_V1(bigistore_exists_any);
+Datum bigistore_exists_any(PG_FUNCTION_ARGS)
+{
+    BigIStorePair *pairs;
+    BigIStore *    is     = PG_GETARG_BIGIS(0);
+    ArrayType *    a      = PG_GETARG_ARRAYTYPE_P_COPY(1);
+    int32 *        ar     = (int32 *) ARR_DATA_PTR(a);
+    int            alen   = ArrayGetNItems(ARR_NDIM(a), ARR_DIMS(a));
+    int            index1 = 0, index2 = 0;
+
+    if (ARR_HASNULL(a) && array_contains_nulls(a))
+        ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED), errmsg("array must not contain nulls")));
+
+    pairs = FIRST_PAIR(is, BigIStorePair);
+
+    if (alen > 1)
+        qsort((void *) ar, alen, sizeof(int32), is_int32_arr_comp);
+
+    while (index1 < is->len && index2 < alen)
+    {
+        if (pairs[index1].key < ar[index2])
+        {
+            ++index1;
+        }
+        else if (pairs[index1].key > ar[index2])
+        {
+            ++index2;
+        }
+        else
+            PG_RETURN_BOOL(true);
+    }
+
+    PG_RETURN_BOOL(false);
+}
+
+PG_FUNCTION_INFO_V1(bigistore_exists_all);
+Datum bigistore_exists_all(PG_FUNCTION_ARGS)
+{
+    BigIStorePair *pairs;
+    BigIStore *    is     = PG_GETARG_BIGIS(0);
+    ArrayType *    a      = PG_GETARG_ARRAYTYPE_P_COPY(1);
+    int32 *        ar     = (int32 *) ARR_DATA_PTR(a);
+    int            alen   = ArrayGetNItems(ARR_NDIM(a), ARR_DIMS(a));
+    int            index1 = 0, index2 = 0, found = 0;
+
+    if (ARR_HASNULL(a) && array_contains_nulls(a))
+        ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED), errmsg("array must not contain nulls")));
+
+    pairs = FIRST_PAIR(is, BigIStorePair);
+
+    if (alen > 1)
+        qsort((void *) ar, alen, sizeof(int32), is_int32_arr_comp);
```

Comment:
Sorting the array + full iteration over an istore is something like `O(N*log(N) + M + N)` on average and `O(N*N + M + N + N)` in the worst case, while it's possible to get more predictable and probably faster `O(N*log(M))` by iterating over an array argument (without copying it) and calling `bigistore_find`. That also applies to the previous function. It might not be a concern though, as both functions are expected to return early.

Reply:
oh well that's a good argument also makes the code a lot simpler

Reply:
So, something like a snippet below is slightly faster in my unscientific test and, IMO, is a bit more straightforward: copying "surviving" elements into a new place in blocks instead of copying everything and then iteratively collapsing it. Although, despite more straightforward logic, it might be even harder to read.

```C
PG_FUNCTION_INFO_V1(istore_delete_array);
Datum istore_delete_array(PG_FUNCTION_ARGS)
{
    IStorePair * pairs, * dst_pairs;
    IStore     * result;
    IStore     * is      = PG_GETARG_IS(0);
    ArrayType  * a       = PG_GETARG_ARRAYTYPE_P_COPY(1);
    int32      * ar      = (int32 *) ARR_DATA_PTR(a);
    int          alen    = ArrayGetNItems(ARR_NDIM(a), ARR_DIMS(a));
    int          is_kept_offset = 0, ar_index = 0, is_left = 0;

    if (ARR_HASNULL(a) && array_contains_nulls(a))
        ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED), errmsg("array must not contain nulls")));

    if (alen > 1)
        qsort((void *) ar, alen, sizeof(int32), is_int32_arr_comp);

    is_left   = is->len;
    pairs     = FIRST_PAIR(is, IStorePair);
    result    = palloc0(ISTORE_SIZE(is, IStorePair));
    SET_VARSIZE(result, ISTORE_SIZE(is, IStorePair));
    dst_pairs = FIRST_PAIR(result, IStorePair);

    while ((is_left - is_kept_offset > 0) && ar_index < alen)
    {
        if (pairs[is_kept_offset].key < ar[ar_index])
        {
            ++is_kept_offset;
        }
        else if (pairs[is_kept_offset].key > ar[ar_index])
        {
            ++ar_index;
        }
        else
        {
            if (is_kept_offset > 0) 
            {
                memcpy(dst_pairs, pairs, is_kept_offset * sizeof(IStorePair));

                dst_pairs   += is_kept_offset;
                result->len += is_kept_offset;

                pairs       += is_kept_offset;
                is_left     -= is_kept_offset;

                is_kept_offset = 0;
            } 

            ++pairs;
            --is_left;
            ++ar_index;
        }
    }
    if (is_left > 0)
    {
        memcpy(dst_pairs, pairs, is_left * sizeof(IStorePair));
        result->len += is_left;
    }

    result = repalloc(result, ISTORE_SIZE(result, IStorePair));
    SET_VARSIZE(result, ISTORE_SIZE(result, IStorePair));

    PG_RETURN_POINTER(result);
}
```

And here is the SQL snippet I used for speed comparison:
```SQL
with
 keys as (
	select s1 as id, array_agg(s2) as ks 
	from generate_series(1, 1000) s1, generate_series(3, 1000, 6) s2
	group by s1
),
 istores as (
	select s1 as id, sum(istore(ARRAY[s1, s2])) as ist
	from generate_series(1, 1000) s1, generate_series(0, 2000, 3) s2
	group by s1
)
select max(istore_length(delete(ist, ks))) 
from istores join keys on istores.id = keys.id;
```